src/messenger/messenger.cpp

```diff
@@ -0,0 +1,242 @@
+ *   3. ...
+ *
+ * To find out more about ZeroMQ, the best resource I've found is the official book: http://zguide.zeromq.org/
+ */
+
+namespace terrier::messenger {
+
+/** An abstraction around ZeroMQ messages which explicitly have the sender specified. */
+class ZmqMessage {
+ public:
+  /** The routing ID of the message sender. */
+  std::string identity_;
+  /** The payload in the message. */
+  std::string payload_;
+};
+
+}  // namespace terrier::messenger
+
+namespace terrier {
+
+/**
+ * Useful ZeroMQ utility functions implemented in a naive manner. Most functions have wasteful copies.
+ * If performance indicates that these functions are a bottleneck, switch to the zero-copy messages of ZeroMQ.
+ */
+class ZmqUtil {
+ public:
+  /** Static utility class. */
+  ZmqUtil() = delete;
+
+  /** The maximum length of a routing ID. Specified by ZeroMQ. */
+  static constexpr int MAX_ROUTING_ID_LEN = 255;
+
+  /** @return The routing ID of the socket. */
+  static std::string GetRoutingId(zmq::socket_t *socket) {
+    char buf[MAX_ROUTING_ID_LEN];
+    size_t routing_id_len;
+    socket->getsockopt(ZMQ_ROUTING_ID, &buf, &routing_id_len);
+    return std::string(buf, routing_id_len);
+  }
+
+  /** @return The next string read off the socket. */
+  static std::string Recv(zmq::socket_t *socket, zmq::recv_flags flags) {
+    zmq::message_t message;
+    auto received = socket->recv(message, flags);
+    if (!received.has_value()) {
+      throw MESSENGER_EXCEPTION(fmt::format("Unable to receive on socket: {}", ZmqUtil::GetRoutingId(socket)));
+    }
+    return std::string(static_cast<char *>(message.data()), received.value());
+  }
+
+  /** @return The next ZmqMessage (identity and payload) read off the socket. */
+  static messenger::ZmqMessage RecvMsg(zmq::socket_t *socket) {
+    std::string identity = Recv(socket, zmq::recv_flags::none);
+    std::string delimiter = Recv(socket, zmq::recv_flags::none);
+    std::string payload = Recv(socket, zmq::recv_flags::none);
+
+    return messenger::ZmqMessage{identity, payload};
+  }
+
+  /** @return Send the specified ZmqMessage (identity and payload) over the socket. */
+  static void SendMsg(zmq::socket_t *socket, const messenger::ZmqMessage &msg) {
+    zmq::message_t identity_msg(msg.identity_.data(), msg.identity_.size());
+    zmq::message_t delimiter_msg("", 0);
+    zmq::message_t payload_msg(msg.payload_.data(), msg.payload_.size());
+    bool ok = true;
+
+    ok = ok && socket->send(identity_msg, zmq::send_flags::sndmore);
+    ok = ok && socket->send(delimiter_msg, zmq::send_flags::sndmore);
+    ok = ok && socket->send(payload_msg, zmq::send_flags::none);
+
+    if (!ok) {
+      throw MESSENGER_EXCEPTION(fmt::format("Unable to send on socket: {}", ZmqUtil::GetRoutingId(socket)));
+    }
+  }
+};
+
+}  // namespace terrier
+
+namespace terrier::messenger {
+
+/** An abstraction around successful connections made through ZeroMQ. */
+class ConnectionId {
```

Comment:
This also needs to be put into header right? I am getting incomplete type error whenI tried to obtain a conn id.

Reply:
You are right. I have pushed an update.